inventory.py

```diff
@@ -1,97 +1,128 @@
+            order_quantity = get_valid_product_quantity()
+            order.add_purchase(order_action, order_quantity)
+            print(f'Added product: {order_action}, quantity: {order_quantity}')
+
 def main():
     main_inventory = Inventory({'apple': 10, 'banana': 10, 'orange': 10})
 
     while True:
-        user_action = input('What would you like to do?\n1. Add Product\n2. Remove Product\n3. Check Stock\n4. Create Order\n5. Exit\n')
+        # Instead of having big long strings here with \n characters, you can actually use
+        # triple quotes and Python will format it exactly as you write it
+        initial_prompt = """
+            What would you like to do?
+            1. Add Product
+            2. Remove Product
+            3. Check Stock
+            4. Create Order
+            5. Exit
+        """
+        # You need to make sure to use the textwrap.dedent() function to remove the
+        # white space at the beginning. This triple quote stuff is more a "nice to have"
+        # than something I would expect everyone to know how to do
+        initial_prompt_formatted = textwrap.dedent(initial_prompt)
+        user_action = input(initial_prompt_formatted)
 
         # input validation for user_action
         if user_action not in Inventory.ALLOWED_ACTIONS:
             print('Invalid Action\n')
         elif user_action == '1':
             product_name = input('Please provide the name of the product you would like to add\n').lower()
-            # input validation for product_quantity 
-            while True:
-                product_quantity = input('Please provide the product quantity\n')
-                if product_quantity.isdigit() == False:
-                    print('Product quantity must be a valid number.\n')
-                elif product_quantity.isdigit() == True:
-                    product_quantity = int(product_quantity)
-                    break
+            # These nested loops are making it hard to read. Move them into their own functions
+            product_quantity = get_valid_product_quantity()
             main_inventory.add_product(product_name, product_quantity)
         elif user_action == '2':
             product_name = input('Please provide the name of the product you would like to remove\n').lower()
             # input validation for product_quantity 
-            while True:
-                product_quantity = input('Please provide the product quantity\n')
-                if product_quantity.isdigit() == False:
-                    print('Product quantity must be a valid number.\n')
-                elif product_quantity.isdigit() == True:
-                    product_quantity = int(product_quantity)
-                    break
-            main_inventory.remove_product(product_name, product_quantity)
+            # Keep in mind the DRY principle: Don't Repeat Yourself. If you find yourself copying and pasting
+            # code, it should probably be its own function

```

Comment:
If you know any other principles like this, they're really helpful for me.

Reply:
https://en.wikipedia.org/wiki/SOLID